### PR TITLE
Generates RPCs as SAM

### PR DESCRIPTION
**Pull request by oldergod**
#957

### CODE REVIEW

File: wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt [hunk 1]
```diff
@@ -112,7 +113,10 @@ sealed class Target {
     val android: Boolean = false,
 
     /** True for emitted types to implement APIs for easier migration from the Java target. */
-    val javaInterop: Boolean = false
+    val javaInterop: Boolean = false,
+
+    /** True for emitted services to implement one interface per RPC. */
+    val servicesAsSingleMethod: Boolean = false
```
Comment (oldergod) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Target.kt]: `serviceAsSingleMethod`?
`serviceAsSAM`?
`serviceAsSingleMethods`?

File: wire-test-utils/src/main/java/com/squareup/wire/schema/RepoBuilder.java [hunk 2]
```diff
@@ -133,4 +133,17 @@ public String generateGrpcKotlin(String serviceName) {
         .build();
     return fileSpec.toString();
   }
+
+  public String generateGrpcKotlinAsSingleMethod(String serviceName, String rpcName) {
```
Comment (swankjesse) [inline wire-test-utils/src/main/java/com/squareup/wire/schema/RepoBuilder.java]: we probably want to make this file Kotlin and use default parameters

File: wire-kotlin-generator/src/test/java/com/squareup/wire/kotlin/KotlinGeneratorTest.kt [hunk 3]
```diff
@@ -328,6 +328,61 @@ class KotlinGeneratorTest {
...
+          |import com.squareup.wire.Service
+          |import com.squareup.wire.WireRpc
+          |
+          |interface RouteGuideGetFeature : Service {
+          |    @WireRpc(
+          |            path = "/routeguide.RouteGuide/GetFeature",
+          |            requestAdapter = "routeguide.Point#ADAPTER",
+          |            responseAdapter = "routeguide.Feature#ADAPTER"
+          |    )
+          |    suspend fun GetFeature(request: Point): Feature
```
Comment (swankjesse) [inline wire-kotlin-generator/src/test/java/com/squareup/wire/kotlin/KotlinGeneratorTest.kt]: this looks good

I wanna have a blocking option somehow

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 4]
```diff
@@ -109,6 +108,17 @@ class KotlinGenerator private constructor(
         .build()
   }
 
+  fun generateServiceAsSingleMethod(service: Service, rpc: Rpc): TypeSpec {
+    val interfaceName = service.name() + rpc.name()
+    val superinterface = com.squareup.wire.Service::class.java
```
Comment (Egorand) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: Use an import alias if there's a name clash

Comment (oldergod) [reply wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: Clash with something named `Service`?

File: wire-kotlin-generator/src/test/java/com/squareup/wire/kotlin/KotlinGeneratorTest.kt (hunk 3 above)
Comment (oldergod) [reply wire-kotlin-generator/src/test/java/com/squareup/wire/kotlin/KotlinGeneratorTest.kt]: it can come later I guess

**Pull request merged**